lib/stream.js

```diff
@@ -0,0 +1,77 @@
+module.exports = Stream
+
+var inherits = require('inherits')
+var stream = require('readable-stream')
+
+inherits(Stream, stream.Readable)
+
+function Stream (chunkstore, opts) {
+  if (!(this instanceof Stream)) return new Stream(chunkstore, opts)
+  if (!opts) opts = {}
+
+  stream.Readable.call(this)
+
+  this.chunkstore = chunkstore
+  this.start = opts.start || 0
+  this.end = opts.end || chunkstore.length
+  this.destroyed = false
+
+  this._startIndex = Math.floor(this.start / chunkstore.chunkLength)
+  this._endIndex = Math.floor(this.end / chunkstore.chunkLength)
+  this._index = this._startIndex
+  this._missing = {}
+
+  if (typeof this.end === 'undefined') {
+    throw new Error('Must define opts.end or chunkstore.length')
+  }
+
+  this.chunkstore._store.on('set', this._onPiece.bind(this))
+}
+
+Stream.prototype._read = function () {
+  var self = this
+
+  if (self._index > self._endIndex) return self.push(null)
+
+  self._get(self._index, function (err, chunk) {
+    if (err) return self._destroy(err)
+    var start = self._index !== self._startIndex ? 0
```

Comment:
A comment here would be helpful. It took me quite a while to understand why we need to calculate the remainder if the index is at the start or end of the stream